sql: enum metadata unset

See sqlsmith/setup=seed/setting=no-mutations: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_WorkloadNightly/2080704?buildTab=artifacts

```		stmt:
		WITH
			with_401 (col_2660)
				AS (
					SELECT * FROM (VALUES ('good day':::public.greeting), ('howdy':::public.greeting)) AS tab_1001 (col_2660)
					INTERSECT ALL
						SELECT
							NULL AS col_2661
						FROM
							defaultdb.public.seed@seed__int8__float8__date_idx AS tab_1002,
							defaultdb.public.seed@seed__int8__float8__date_idx AS tab_1003
				),
			with_402 (col_2662) AS (SELECT NULL AS col_2662)
		SELECT
			cte_ref_110.col_2660 AS col_2663
		FROM
			with_401 AS cte_ref_110
		ORDER BY
			cte_ref_110.col_2660 ASC, cte_ref_110.col_2660, cte_ref_110.col_2660 DESC;
```

Failed with:

```
panic: use of enum metadata before hydration as an enum

goroutine 3630 [running]:
panic(0x469dee0, 0xc001afd1e0)
	/usr/local/go/src/runtime/panic.go:1064 +0x46d fp=0xc0020a31f0 sp=0xc0020a3138 pc=0x7afccd
github.com/cockroachdb/cockroach/pkg/sql/types.(*T).ensureHydratedEnum(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/types/types.go:2320
github.com/cockroachdb/cockroach/pkg/sql/types.(*T).EnumGetIdxOfPhysical(0xc00054db80, 0xc00179b07d, 0x1, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/types/types.go:2287 +0x33b fp=0xc0020a32b8 sp=0xc0020a31f0 pc=0x1344c0b
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.GetEnumComponentsFromPhysicalRep(0xc00054db80, 0xc00179b07d, 0x1, 0x1, 0x0, 0x0, 0xc00179b07d, 0x1, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:3813 +0x4d fp=0xc0020a3300 sp=0xc0020a32b8 pc=0x15138cd
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DecodeUntaggedDatum(0xc00163fb38, 0xc00054db80, 0xc00179b07c, 0x2, 0x2, 0x0, 0x6, 0x0, 0x0, 0x78e42e, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/column_type_encoding.go:630 +0x2e6c fp=0xc0020a3930 sp=0xc0020a3300 pc=0x170aedc
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DecodeTableValue(0xc00163fb38, 0xc00054db80, 0xc00179b07b, 0x3, 0x3, 0x0, 0x8, 0x556c0e0, 0xc00163f800, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/column_type_encoding.go:496 +0x160 fp=0xc0020a39a0 sp=0xc0020a3930 pc=0x1707f80
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*EncDatum).EnsureDecoded(0xc001488098, 0xc00054db80, 0xc00163fb38, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:232 +0xa0 fp=0xc0020a3a90 sp=0xc0020a39a0 pc=0x171a8b0
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*EncDatum).Encode(0xc001488098, 0xc00054db80, 0xc00163fb38, 0xc000000000, 0xc000c65200, 0x0, 0x8, 0xc00140e000, 0x0, 0xc0018998c0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:267 +0x165 fp=0xc0020a3b40 sp=0xc0020a3a90 pc=0x171af85
github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.encodeColumnsOfRow(0xc00163fb38, 0xc000c65200, 0x0, 0x8, 0xc001488098, 0x1, 0x1, 0xc000c1e5f0, 0x1, 0x14, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/hash_row_container.go:129 +0x173 fp=0xc0020a3bc0 sp=0xc0020a3b40 pc=0x19956d3
github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*columnEncoder).encodeEqualityCols(0xc00163fb08, 0x54e1e80, 0xc0018a2440, 0xc001488098, 0x1, 0x1, 0xc000c1e5f0, 0x1, 0x14, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/hash_row_container.go:143 +0xd4 fp=0xc0020a3c88 sp=0xc0020a3bc0 pc=0x1995894
github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*hashMemRowBucketIterator).Reset(0xc001bc7f80, 0x54e1e80, 0xc0018a2440, 0xc001488098, 0x1, 0x1, 0xc000c1e500, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/hash_row_container.go:360 +0x97 fp=0xc0020a3d08 sp=0xc0020a3c88 pc=0x19967a7
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*hashJoiner).readProbeSide(0xc00140d000, 0x3, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/hashjoiner.go:467 +0x15c fp=0xc0020a3dc0 sp=0xc0020a3d08 pc=0x1cf10cc
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*hashJoiner).Next(0xc00140d000, 0x7, 0x1, 0x54e1e80, 0xc0018a2440)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/hashjoiner.go:237 +0x27f fp=0xc0020a3e28 sp=0xc0020a3dc0 pc=0x1cf022f
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sortAllProcessor).fill(0xc000858f00, 0x54e1e80, 0xc0018a2440, 0x4a3281b)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sorter.go:291 +0x73 fp=0xc0020a3ef8 sp=0xc0020a3e28 pc=0x1d1d493
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*sortAllProcessor).Start(0xc000858f00, 0x54e1e80, 0xc0018a2440, 0xc0002465a0, 0x54e1e80)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/sorter.go:272 +0xab fp=0xc0020a3f40 sp=0xc0020a3ef8 pc=0x1d1d3bb
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run(0xc000858f00, 0x54e1e80, 0xc0018a2440)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:747 +0x52 fp=0xc0020a3f80 sp=0xc0020a3f40 pc=0x1ac14e2
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal.func1(0xc001f749b0, 0x2, 0x5, 0x54e1e80, 0xc0018a2440, 0xc0017bb440, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:315 +0x5c fp=0xc0020a3fa8 sp=0xc0020a3f80 pc=0x1cdef2c
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc0020a3fb0 sp=0xc0020a3fa8 pc=0x7e3281
created by github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:314 +0x2d9
```

--- a/pkg/sql/logictest/testdata/logic_test/enums
+++ b/pkg/sql/logictest/testdata/logic_test/enums
@@ -1062,3 +1062,14 @@ query T
 SELECT to_json('hello'::greeting)
 ----
 "hello"
+
+# Regression for #51474.
+statement ok
+CREATE TABLE t51474 (x greeting);
+INSERT INTO t51474 VALUES ('hello'), ('howdy')
+
+query T rowsort
+SELECT * FROM t51474 INTERSECT ALL SELECT * FROM t51474
+----
+hello
+howdy

--- a/pkg/sql/rowflow/row_based_flow.go
+++ b/pkg/sql/rowflow/row_based_flow.go
@@ -268,6 +268,15 @@ func (f *rowBasedFlow) setupInputSyncs(
 				return nil, errors.Errorf("unsupported input sync type %s", is.Type)
 			}
 
+			// Before we can safely use types we received over the wire in the
+			// processors, we need to make sure they are hydrated. All processors other
+			// than processors that scan over tables get their inputs from here, so
+			// this is a convenient place to do the hydration. Processors that scan
+			// over tables will have their hydration performed in ProcessorBase.Init.
+			if err := execinfrapb.HydrateTypeSlice(f.EvalCtx, is.ColumnTypes); err != nil {
+				return nil, err
+			}
+
 			if is.Type == execinfrapb.InputSyncSpec_UNORDERED {
 				if opt == flowinfra.FuseNormally || len(is.Streams) == 1 {
 					// Unordered synchronizer: create a RowChannel for each input.

